pkg/database/user_report.go

```diff
@@ -110,13 +120,74 @@ func (db *Database) FindUserReport(tx *gorm.DB, phoneNumber string) (*UserReport
 
 // DeleteUserReport removes a specific phone number from the user report
 // de-duplication table.
-func (db *Database) DeleteUserReport(phoneNumber string) error {
+// The actor may be nil on this operation, if it's the result of a
+// natural delete (SMS failure).
+// Otherwise this is an administrative acction.
+func (db *Database) DeleteUserReport(phoneNumber string, actor Auditable) error {
 	hmacedCodes, err := db.generatePhoneNumberHMACs(phoneNumber)
 	if err != nil {
 		return fmt.Errorf("failed to create hmac: %w", err)
 	}
 
-	return db.db.Where("phone_hash IN (?)", hmacedCodes).Delete(&UserReport{}).Error
+	return db.db.Transaction(func(tx *gorm.DB) error {
+		var ur UserReport
+		if err := tx.
+			Set("gorm:query_option", "FOR UPDATE").
+			Model(&UserReport{}).
+			Where("phone_hash IN (?)", hmacedCodes).
+			First(&ur).
+			Error; err != nil {
+			if IsNotFound(err) {
+				// Log an audit record on attempted purges as well.
+				if actor != nil {
```

Comment:
I'm not entirely convinced this makes sense. Deleting a phone number is already a highly-privileged operation. I don't think we need to generate an audit log when deleting the record failed due to an invalid phone number. If we're worried about abuse, let's put a rate limit on the number of times this API can be called in a minute/hour.

Reply:
but it's not...
the two existing cases 
- user report fails at the SMS stage
- e2e runner

I don't want to log those two cases.

Reply:
I'm suggesting we _don't_ generate an audit when deleting a user report if the user report is not found. I'm advocating we delete L142-153 entirely. The audit log should really only contain actions that mutated data (and this would not have mutated data since the data didn't exist).

Reply:
we should discuss offline. This is to catch an admin attempting to delete a bunch of phone numbers even if they weren't there.